File: backend/app/models/daily_care.py
from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from app.models.vaccination import utc_now

"""Today's routine wellness care — the small daily nudges on the dashboard.

One document per dog per DAY. The date is part of the key rather than
something a scheduled job clears, so "resets at midnight" needs no cron and
cannot half-fail: tomorrow simply reads a different key and finds nothing
ticked. It also means the history is kept for free, which is what would make
a "fed 6 days out of 7" summary possible later.

Deliberately NOT medical. These are habit nudges (fed, water, walk, groom);
vaccinations, dewormings and prescriptions stay in their own collections
where the evidence rules and the review flow live.
"""

#: The four toggles the dashboard shows, in display order. Codes are stored;
#: labels live here so a rewording can never split one habit into two rows in
#: whatever history we build on top of this.
CARE_ITEMS = ("fed", "water", "walk", "grooming")

CARE_LABELS = {
    "fed": "Fed",
    "water": "Fresh water",
    "walk": "Walked",
    "grooming": "Grooming",
}

CARE_ICONS = {
    "fed": "🍚",
    "water": "💧",
    "walk": "🦮",
    "grooming": "🧴",
}


class DailyCare(BaseModel):
    """One dog's routine-care ticks for one calendar day."""

    id: str
    dog_id: str
    #: ISO date (YYYY-MM-DD) in the SERVER's day, matching how every other
    #: due-date in the app is computed. A device with a wrong clock can't
    #: silently tick tomorrow's box.
    day: str
    #: item code -> when it was ticked. Absent means not done; the timestamp
    #: is kept rather than a bare bool so "fed at 8am" is answerable later.
    done: Dict[str, datetime] = Field(default_factory=dict)
    updated_at: datetime = Field(default_factory=utc_now)


class DailyCareToggle(BaseModel):
    item: str
    done: bool
    #: Optional: lets a client be explicit rather than relying on the server's
    #: idea of "today" when the two could disagree around midnight.
    day: Optional[str] = None


class CareItem(BaseModel):
    """One habit on the routine card. `code` is the stable key the daily ticks
    are recorded against — renaming an item MUST keep its code, or the day's
    history silently detaches from it.

    Optional because a NEW item has no code yet; the server assigns one. An
    existing item must send its code back, which is exactly what preserves
    today's ticks across a rename."""

    code: Optional[str] = None
    label: str
    icon: str = "🐾"


class CareItemList(BaseModel):
    """The whole list, replaced in one call. Sending the complete list rather
    than per-item edits means the order on screen is exactly the order stored,
    with no separate reordering endpoint to keep in sync."""

    items: List[CareItem]

File: frontend/src/utils/dailyCare.js
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

/* Today's routine wellness care.

   The server owns both the item list and the "today" boundary, so the card
   can't drift from it: a device with a wrong clock still ticks the server's
   day, and adding a fifth habit later needs no frontend change. */

export async function fetchDailyCare(dogId) {
  if (!dogId) return null;
  try {
    const { data } = await axios.get(`${API_URL}/api/daily-care/${dogId}`);
    return data;
  } catch {
    return null;   // the card hides rather than showing a broken shell
  }
}

export async function toggleDailyCare(dogId, item, done) {
  const { data } = await axios.patch(`${API_URL}/api/daily-care/${dogId}`, { item, done });
  return data;
}

/* Replace the whole list in one call.

   An existing item MUST send its `code` back — that is what keeps today's
   tick attached to it through a rename. A new item sends no code and the
   server assigns one. */
export async function saveCareItems(dogId, items) {
  const { data } = await axios.put(`${API_URL}/api/daily-care/${dogId}/items`, {
    items: items.map(({ code, label, icon }) => ({ code: code || undefined, label, icon })),
  });
  return data;   // the card's full refreshed state
}

/** A small, deliberately boring set — enough to tell habits apart at a glance. */
export const ICON_CHOICES = [
  '🍚', '💧', '🦮', '🧴', '💊', '🦷', '🎾', '🛁', '✂️', '🧠', '🐾', '❤️',
];

/** "Resets in 7 hours" / "Resets in 45 minutes" — the card's own countdown. */
export const resetLabel = (seconds) => {
  if (seconds == null) return '';
  const hours = Math.floor(seconds / 3600);
  if (hours >= 1) return `Resets in ${hours} hour${hours === 1 ? '' : 's'}`;
  const minutes = Math.max(1, Math.round(seconds / 60));
  return `Resets in ${minutes} minute${minutes === 1 ? '' : 's'}`;
};
